lib/std/zig/system.zig

```diff
@@ -461,139 +444,127 @@ pub fn abiAndDynamicLinkerFromFile(
-        // even when the ELF file is 32-bits.
-        const ph_reserve: usize = @sizeOf(elf.Elf64_Phdr) - @sizeOf(elf.Elf32_Phdr);
-        const ph_read_byte_len = try preadMin(file, ph_buf[0 .. ph_buf.len - ph_reserve], phoff, phentsize);
-        var ph_buf_i: usize = 0;
-        while (ph_buf_i < ph_read_byte_len and ph_i < phnum) : ({
-            ph_i += 1;
-            phoff += phentsize;
-            ph_buf_i += phentsize;
-        }) {
-            const ph32: *elf.Elf32_Phdr = @ptrCast(@alignCast(&ph_buf[ph_buf_i]));
-            const ph64: *elf.Elf64_Phdr = @ptrCast(@alignCast(&ph_buf[ph_buf_i]));
-            const p_type = elfInt(is_64, need_bswap, ph32.p_type, ph64.p_type);
-            switch (p_type) {
-                elf.PT_INTERP => if (look_for_ld) {
-                    const p_offset = elfInt(is_64, need_bswap, ph32.p_offset, ph64.p_offset);
-                    const p_filesz = elfInt(is_64, need_bswap, ph32.p_filesz, ph64.p_filesz);
-                    if (p_filesz > result.dynamic_linker.buffer.len) return error.NameTooLong;
-                    const filesz = @as(usize, @intCast(p_filesz));
-                    _ = try preadMin(file, result.dynamic_linker.buffer[0..filesz], p_offset, filesz);
-                    // PT_INTERP includes a null byte in filesz.
-                    const len = filesz - 1;
-                    // dynamic_linker.max_byte is "max", not "len".
-                    // We know it will fit in u8 because we check against dynamic_linker.buffer.len above.
-                    result.dynamic_linker.max_byte = @as(u8, @intCast(len - 1));
-
-                    // Use it to determine ABI.
-                    const full_ld_path = result.dynamic_linker.buffer[0..len];
-                    for (ld_info_list) |ld_info| {
-                        const standard_ld_basename = fs.path.basename(ld_info.ld.get().?);
-                        if (std.mem.endsWith(u8, full_ld_path, standard_ld_basename)) {
-                            result.abi = ld_info.abi;
-                            break;
-                        }
+    if (header.phentsize > @sizeOf(elf.Elf64_Phdr)) return error.InvalidElfFile;
+
+    var got_dyn_section: bool = false;
+    var program_headers = header.program_header_iterator(&file);
+    while (program_headers.next() catch |err| switch (err) {
+        error.OperationAborted => unreachable, // windows-only
+        error.WouldBlock => unreachable,
+        error.NotOpenForReading => unreachable,
+        error.SystemResources, error.Unexpected => |e| return e,
+        error.InputOutput => return error.FileSystem,
+        else => return error.UnableToReadElfFile,
+    }) |*program_header| switch (program_header.p_type) {
+        elf.PT_INTERP => {
+            got_dyn_section = true;
+            if (look_for_ld) {
+                const p_offset = program_header.p_offset;
+                const p_filesz = program_header.p_filesz;
```

Comment:
The code that you are replacing here is designed to make as few syscalls as possible. Can you share how the new code fared against the old code?